node-gateway.js

```diff
@@ -63,17 +60,32 @@ class NodeGateway extends ApolloGateway {
     const modules = [];
     const seenNodeTypes = new Set();
     for (const service of defs.serviceDefinitions) {
-      visit(service.typeDefs, {
+      // Manipulate the typeDefs of the service
+      service.typeDefs = visit(service.typeDefs, {
         ObjectTypeDefinition(node) {
           const name = node.name.value;
+
+          // Remove existing `query { node }` from service
+          if (name === 'Query') {
+            return visit(node, {
+              FieldDefinition(node) {
+                const name = node.name.value;
+                if (name === 'node') {
+                  return null;
+                }
+              },
+            });
+          }
+
+          // Find the the types that implement the Node interface
           if (!isNode(node) || seenNodeTypes.has(name)) {
             return;
+          } else {
+            // We don't need any resolvers for these modules; they're just
+            // simple objects with a single `id` property.
+            modules.push({ typeDefs: toTypeDefs(name) });
+            seenNodeTypes.add(name);
           }
-
-          // We don't need any resolvers for these modules; they're just
-          // simple objects with a single `id` property.
-          modules.push({ typeDefs: toTypeDefs(name) });
-          seenNodeTypes.add(name);
```

Comment:
You don't need an `else` after an `if` that you've returned in. I'd suggest inverting the second condition so the branches are on "equal footing". Something like this?

```typescript
service.typeDefs = visit(service.typeDefs, {
  ObjectTypeDefinition(node) {
    const name = node.name.value;

    // Remove existing `query { node }` from service to avoid collisions
    if (name === 'Query') {
      return visit(node, {
        FieldDefinition(node) {
          if (node.name.value === 'node') {
            return null;
          }
        },
      });
    }

    // Add any new Nodes from this service to the Node service's modules
    if (isNode(node) && !seenNodeTypes.has(name)) {
      // We don't need any resolvers for these modules; they're just
      // simple objects with a single `id` property.
      modules.push({ typeDefs: toTypeDefs(name) });
      seenNodeTypes.add(name);

      return;
    }
  },
});
```

Reply:
I like it! Added 😄